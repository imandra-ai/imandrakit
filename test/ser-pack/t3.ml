[@@@warning "-60"]

module M1 = struct
  Format.printf "without hashcons@."

  type t0 = {
    x: int; [@key "X"]
    y: bool; [@key "Y"]
  }
  [@@deriving serpack, show]

  type t1 = { foos: t0 list } [@@deriving serpack, show]

  let myt1 =
    let x1 = { x = 0; y = false } in
    let x2 = { x = 2; y = true } in
    { foos = [ x1; x1; x2; x1; x2 ] }
  ;;

  Format.printf "myt1: %a@." pp_t1 myt1

  let pack1 = Imandrakit_ser_pack.to_value t1_to_serpack myt1;;

  Format.printf "pack1: %a@." Imandrakit_ser_pack.pp_diagnostic pack1

  let len1 =
    Imandrakit_ser_pack.to_cbor_string t1_to_serpack myt1 |> String.length
  ;;

  Format.printf "len1: %d@." len1
end

module M2 = struct
  Format.printf "with hashcons@."

  type t0 = {
    x: int; [@key "x0"]
    y: bool; [@key "y"]
  }
  [@@deriving serpack, show] [@@serpack.hashcons]

  type t1 = { foos: t0 list } [@@deriving serpack, show]

  let myt1 =
    let x1 = { x = 0; y = false } in
    let x2 = { x = 2; y = true } in
    { foos = [ x1; x1; x2; x1; x2 ] }
  ;;

  Format.printf "myt1': %a@." pp_t1 myt1

  let pack1 = Imandrakit_ser_pack.to_value t1_to_serpack myt1;;

  Format.printf "pack1: %a@." Imandrakit_ser_pack.pp_diagnostic pack1

  let len1 =
    Imandrakit_ser_pack.to_cbor_string t1_to_serpack myt1 |> String.length
  ;;

  Format.printf "len1: %d@." len1
end
