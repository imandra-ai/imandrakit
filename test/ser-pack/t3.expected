without hashcons
myt1: { T3.M1.foos =
        [{ T3.M1.x = 0; y = false }; { T3.M1.x = 0; y = false };
          { T3.M1.x = 2; y = true }; { T3.M1.x = 0; y = false };
          { T3.M1.x = 2; y = true }]
        }
pack1: {h=[0: {"X": 0, "Y": false};
           1: {"X": 0, "Y": false};
           2: {"X": 2, "Y": true};
           3: {"X": 0, "Y": false};
           4: {"X": 2, "Y": true};
           5: {0: [6(0), 6(1), 6(2), 6(3), 6(4)]};
           ];
        k=6(5)}
len1: 56
with hashcons
myt1': { T3.M2.foos =
         [{ T3.M2.x = 0; y = false }; { T3.M2.x = 0; y = false };
           { T3.M2.x = 2; y = true }; { T3.M2.x = 0; y = false };
           { T3.M2.x = 2; y = true }]
         }
pack1: {h=[0: {"x0": 0, "y": false};
           1: {"x0": 2, "y": true};
           2: {0: [6(0), 6(0), 6(1), 6(0), 6(1)]};
           ];
        k=6(2)}
len1: 37
